aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java

```diff
@@ -0,0 +1,157 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.storage.sqlite;
+
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.ModelAssociation;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.core.model.ModelSchemaRegistry;
+import com.amplifyframework.core.model.query.QueryOptions;
+import com.amplifyframework.core.model.query.Where;
+import com.amplifyframework.core.model.query.predicate.QueryField;
+import com.amplifyframework.core.model.query.predicate.QueryPredicate;
+import com.amplifyframework.core.model.query.predicate.QueryPredicateOperation;
+import com.amplifyframework.core.model.query.predicate.QueryPredicates;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.sqlite.adapter.SQLiteTable;
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class SQLiteModelTree {
+    private final ModelSchemaRegistry registry;
+    private final SQLCommandFactory commandFactory;
+    private final SQLiteDatabase database;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param commandFactory SQL command factory
+     * @param database SQLite database connection handle
+     */
+    SQLiteModelTree(ModelSchemaRegistry registry,
+                    SQLCommandFactory commandFactory,
+                    SQLiteDatabase database) {
+        this.registry = registry;
+        this.commandFactory = commandFactory;
+        this.database = database;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
+        if (Empty.check(root)) {
+            throw new IllegalArgumentException("Cannot traverse tree from an empty root.");
+        }
+        Map<ModelSchema, Set<String>> descendants = new LinkedHashMap<>();
```

Comment:
Does this need to be a `LinkedHashMap`, or would a regular `HashMap` work the same way?

Reply:
Ah, I used to think the order matters, but later realized that the order does not matter. I don't think it's a big deal anyways, and it _is_ nice that the map happens to present the entries in the order of generation.

Reply:
Fair point.  The fact that it is ordered makes it deterministic, which might make unit tests easier to write.